### DISCUSSION TITLE
Use case for endian-specific floats f16le, f16be etc

### DISCUSSION THREAD

**Discussion by gitlost** [odin-lang/Odin#1601] [Proposals]
Was wondering what the use case for endian-specific floats was, as currently they're broken (for `f16` and `f32` anyway). E.g. on a little-endian machine, the following

```odin
package check_float_endian

import "core:fmt"

main :: proc() {
	f :f16be = 1

	fmt.printf("f16be %v (%h)\n", f, f);

	f2 :f32be = 1

	fmt.printf("f32be %v (%h)\n", f2, f2);
}
```
will print
```
f16be 0.000 (0h0)
f32be 0.000 (0h0)
```

Of course this can be fixed but if they're not being used much (which them being broken sort of hints at) then removing them from the language would be a very nice simplification.

**Comment by Kelimion** [comment]
This might be a bug in `fmt`?
Does `fmt.printf("0x%04x\n", transmute(u32)f2)` look like it should for that value as a big endian float?

We had a problem with u32be initially as well, but it's exercised as part of the PNG test suite decoding the `gAMA` chunk and testing the value, so that got fixed.

**Comment by gitlost** [comment]
Hi, it might be, but was looking at the `case ExactValue_Float:` in `lb_const_value` (https://github.com/odin-lang/Odin/blob/master/src/llvm_backend_const.cpp#L579) and noticed it just does a swap of the `f64` value (as opposed to the `case ExactValue_Complex` following which calls `lb_const_f16` etc which swap based on endianness) so I don't think so.

I noticed the use of endian-specific ints alright but the float ones seem much less used. There's also the issue of what endianness means for floats, as "half-endian" formats are also used. There's also the more philosophical point that endianness only comes into play at these input/output situations (and maybe for big int calculations) and it seems to me overkill to have as first-class types, particularly for floats (which are hard enough to get right anyway), and the simplification benefits (of removing them for floats at least) are significant.

**Reply by Kelimion** [reply]
Implementing file formats as I do, I have found the first class integer endian types to be very welcome. They make the structs and enums more self-documenting, instead of guessing which of them have to be byte-swapped when, and on which platforms.

As for endian-specific floats, this is something for @gingerBill to look into, but I'd not write them off just yet.

**Comment by gitlost** [comment]
Ha won't do!
